src/storage/FileResourceStore.ts

```diff
@@ -0,0 +1,458 @@
+import { createReadStream, createWriteStream, promises as fsPromises, Stats } from 'fs';
+import { DataFactory, StreamWriter } from 'n3';
+import { InteractionHandler, RequestAction } from '../util/InteractionHandler';
+import { NamedNode, Quad } from 'rdf-js';
+
+/**
+ * Resource store storing its data in the file system backend.
+ * All requests will throw an {@link NotFoundHttpError} if unknown identifiers get passed.
+ */
+export class FileResourceStore implements ResourceStore {
+  private readonly base: string;
+  private readonly root: string;
+  private readonly interactionHandler: InteractionHandler;
+
+  private readonly rdf = 'http://www.w3.org/1999/02/22-rdf-syntax-ns#';
+  private readonly ldp = 'http://www.w3.org/ns/ldp#';
+  private readonly terms = 'http://purl.org/dc/terms/';
+  private readonly xml = 'http://www.w3.org/2001/XMLSchema#';
+  private readonly stat = 'http://www.w3.org/ns/posix/stat#';
+
+  private readonly predicates = {
+    aType: DataFactory.namedNode(`${this.rdf}type`),
+    modified: DataFactory.namedNode(`${this.terms}modified`),
+    contains: DataFactory.namedNode(`${this.ldp}contains`),
+    mtime: DataFactory.namedNode(`${this.stat}mtime`),
+    size: DataFactory.namedNode(`${this.stat}size`),
+  };
+
+  private readonly objects = {
+    container: DataFactory.namedNode(`${this.ldp}Container`),
+    basicContainer: DataFactory.namedNode(`${this.ldp}BasicContainer`),
+    ldpResource: DataFactory.namedNode(`${this.ldp}Resource`),
+    dateTime: DataFactory.namedNode(`${this.xml}dateTime`),
+  };
+
+  /**
+   * @param base - Will be stripped of all incoming URIs and added to all outgoing ones to find the relative path.
+   * @param root - Root filepath in which the resources and containers will be saved as files and directories.
+   */
+  public constructor(base: string, root: string) {
+    this.base = base;
+    while (this.base.endsWith('/')) {
+      this.base = this.base.slice(0, -1);
+    }
+
+    this.root = root;
+    while (this.root.endsWith('/')) {
+      this.root = this.root.slice(0, -1);
+    }
+
+    this.interactionHandler = new InteractionHandler();
+  }
+
+  /**
+   * Store the incoming data as a file under a file path corresponding to `container.path`,
+   * where slashes correspond to subdirectories.
+   * @param container - The identifier to store the new data under.
+   * @param representation - Data to store. Only File streams are supported.
+   *
+   * @returns The newly generated identifier.
+   */
+  public async addResource(container: ResourceIdentifier, representation: Representation): Promise<ResourceIdentifier> {
+    if (representation.dataType !== DATA_TYPE_BINARY) {
+      throw new UnsupportedMediaTypeHttpError('FileResourceStore only supports binary representations.');
+    }
+    const path = this.parseIdentifier(container);
+    const { slug, linkType, raw } = representation.metadata;
+    let metadata;
+    if (raw && raw.length > 0) {
```

Comment:
`raw?.length > 0`

Reply:
`raw` is not an optional parameter so the first check should not even be necessary.